text/0000-gen-fn.md

```diff
@@ -0,0 +1,785 @@
+
+### `gen` blocks
+
+`gen` blocks are the same as an unstable generator...
+
+* ...without arguments,
+* ...with an additional check forbidding holding borrows across `yield` points,
+* ...and with an automatic `Iterator` implementation.
+
+We'll probably be able to modularize the generator implementation and make it more robust on the implementation and diagnostics side for the `gen` block case, but I believe the initial implementation should be a HIR lowering to a generator and wrapping that generator in [`from_generator`][].
+
+## Fusing
+
+Special case the generator implementation if `Generator::Return = ()` to not panic, but
+repeatedly produce values of the unit type.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+It's another language feature for something that can already be written entirely in user code.
+
+In contrast to `Generator`, `gen` blocks that produce `Iterator`s cannot hold references across `yield` points.
+See [`from_generator`][] which has an `Unpin` bound on the generator it takes to produce an `Iterator`.
+
+[`from_generator`]: https://doc.rust-lang.org/std/iter/fn.from_generator.html
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+## Keyword
+
+We could use `iter` as the keyword.
+I prefer `iter` because I connect generators with a more powerful scheme than plain `Iterator`s.
+The `Generator` trait can do everything that `iter` blocks and `async` blocks can do and more.
+I believe connecting the `Iterator` trait with `iter` blocks is the right choice,
+but that would require us to carve out many exceptions for this keyword as `iter` is used for module names and method names everywhere (including libstd/libcore).
+It may not be much worse than `gen` (see also [the unresolved questions][unresolved-questions]).
+We may want to use `gen` for full on generators in the future.
+
+## Do not do this
+
+One alternative is to keep adding more helper methods to `Iterator`.
+It is already hard for new Rustaceans to be aware of all the capabilities of `Iterator`.
+Some of these new methods would need to be very generic.
+While it's not an `Iterator` example, [`array::try_map`][] is something that has very complex diagnostics that are hard to improve, even if it's nice once it works.
+
+Users can use crates like [`genawaiter`](https://crates.io/crates/genawaiter) or [`propane`](https://crates.io/crates/propane) instead.
+`genawaiter` works on stable and provides `gen!` macro blocks that behave like `gen` blocks, but don't have compiler support for nice diagnostics or language support for the `?` operator. The `propane` crate uses the `Generator` trait from nightly and works mostly
+like `gen` would.
+
+The standard library includes [`std::iter::from_fn`][], which can be used in
+some cases, but as we saw in the example [above][motivation], often the
+improvement over writing out a manual implementation of `Iterator` is limited.
```

Comment:
Well, the example also shows that our existing combinators are better (more concise without being less clear) than generators. I think it'd be good to have an example where `gen` actually beats the status quo.

Reply:
What would be the current way to [write this](https://github.com/estebank/iterator_item/blob/b79d7b496687b1a0f995bf16ca6482f37061dbca/tests/merge_overlapping_intervals.rs#L62-L73) that is more concise than the generator version?

Reply:
I'm sure there's a lot of nice examples. :) I had to write iterator structs myself when I really just wanted `yield`. ([Recent example](https://github.com/rust-lang/rust/blob/f3f9b795bdaccd8284baba295810e87646754c28/compiler/rustc_const_eval/src/interpret/projection.rs#L99-L127).) I'm just saying the RFC should contain such an example.

Your code uses `?` so I'm already unsure what it will even do.^^ I think under this RFC it wouldn't work since `?` on `Option<T>` requires the yield type to be `Option<_>`?

Reply:
We've now updated this example to implement run-length encoding.  This is much more *motivating*.